### ISSUE TITLE
Android : Camera cannot be opened with Cordova 9

### ISSUE THREAD

**Issue by GuiF007** [apache/cordova-plugin-camera#486]
# Bug Report

## Problem
I have an app with a "Open Camera" button. The first time I click on this button, Android requests permission to use Camera.

### What is expected to happen?
After permission asking, I expect the camera opening
When I click on the "Open Camera", the camera opening

### What does actually happen?
After permission asking, the app crashes
When I click on the "Open Camera", nothing happen.

On log, when I click on the button :

> 08-06 17:06:18.679 20453 20602 E PluginManager: Uncaught exception from plugin
> 08-06 17:06:18.679 20453 20602 E PluginManager: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.res.XmlResourceParser android.content.pm.PackageItemInfo.loadXmlMetaData(android.content.pm.PackageManager, java.lang.String)' on a null object reference
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at androidx.core.content.FileProvider.parsePathStrategy(FileProvider.java:605)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at androidx.core.content.FileProvider.getPathStrategy(FileProvider.java:579)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at androidx.core.content.FileProvider.getUriForFile(FileProvider.java:417)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.camera.CameraLauncher.takePicture(CameraLauncher.java:303)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.camera.CameraLauncher.callTakePicture(CameraLauncher.java:282)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.camera.CameraLauncher.execute(CameraLauncher.java:190)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.CordovaPlugin.execute(CordovaPlugin.java:98)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.PluginManager.exec(PluginManager.java:132)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.CordovaBridge.jsExec(CordovaBridge.java:59)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at org.apache.cordova.engine.SystemExposedJsApi.exec(SystemExposedJsApi.java:41)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at android.os.MessageQueue.nativePollOnce(Native Method)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at android.os.MessageQueue.next(MessageQueue.java:379)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at android.os.Looper.loop(Looper.java:144)
> 08-06 17:06:18.679 20453 20602 E PluginManager: 	at android.os.HandlerThread.run(HandlerThread.java:65)

Difference between Cordova 8 and 9, I don't have any provider with Cordova 9. I got the following with Cordova 8 in the AndroidManifest.xml : 

>         <provider android:authorities="${applicationId}.provider" android:exported="false" android:grantUriPermissions="true" android:name="org.apache.cordova.camera.FileProvider">
>             <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/camera_provider_paths" />
>         </provider>

### Version information
- cordova 9.0.0, 
- cordova-android 8.0.0, 
- cordova-plugin-camera 4.0.3 and 4.1.0
- ionic v3
- @ionic-native/camera : ^4.5.2

I DID NOT TEST ON iOS YET.
Works with Cordova 8.1.2 (and cordova-android 7.1.2)

## Checklist

- [X] I searched for existing GitHub issues
- [X] I updated all Cordova tooling to most recent version
- [X] I included all the necessary information above

**Comment by GuiF007** [comment]
I fixed adding this in the config.xml : 

> <config-file mode="merge" parent="application" target="AndroidManifest.xml">
>          [...]
>             <provider android:authorities="${applicationId}.provider" android:exported="false" android:grantUriPermissions="true" android:name="org.apache.cordova.camera.FileProvider">
>                 <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/camera_provider_paths" />
>             </provider>
>         </config-file>

**Comment by LeandroFavero** [comment]
Same problem here, changed config.xml but still not work

**Comment by QuentinIcky** [comment]
Same issue for me, changing config.xml doesn't change... It works perfectly on IOS, but get error 20 (permission) on Android device. Is there any way to fix it ?

**Comment by manpreet1990** [comment]
same problem here , can anyone please help me to resolve that issue. is there any way to fix it ?

**Comment by Mapiac** [comment]
@LeandroFavero @QuentinIcky @manpreet1990 did you guys get this sorted? Same issue.

**Comment by QuentinIcky** [comment]
@Mapiac I didn't solve it because I moved to another project, but just before, I found that if I rollback to an old Android version, it worked. But it's not a good choice, Play Store doesn't accept old Android version.

**Comment by breautek** [comment]
The plugin already contains:

```xml
<provider
  android:name="org.apache.cordova.camera.FileProvider"
  android:authorities="${applicationId}.provider"
  android:exported="false"
  android:grantUriPermissions="true" >
  <meta-data
      android:name="android.support.FILE_PROVIDER_PATHS"
      android:resource="@xml/camera_provider_paths"/>
</provider>
```

And it's been there for a couple of years.

Using cordova 9, I created a new app and added the plugin, and my `AndroidManifest` in the app does contain the mentioned xml that supposedly fix the issue.

This issue was reported more than a month ago, can anyone confirm they still have this issue? Perhaps it was already corrected by a patch in the CLI?

**Comment by jasonandress** [comment]
Yes I still had the problem today. It works after I add this permission

<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />

**Comment by liuxiaoy** [comment]
I meet this. because the plugin config-file
```
        <config-file mode="merge" parent="application" target="AndroidManifest.xml">
            <provider android:authorities="${applicationId}.provider" android:exported="false" android:grantUriPermissions="true" android:name="org.apache.cordova.camera.FileProvider">
                <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/camera_provider_paths" />
            </provider>
        </config-file>
```
 was removed due to a conflict after I add a edit-config
```
<edit-config xmlns:android="http://schemas.android.com/apk/res/android" file="AndroidManifest.xml" mode="merge" target="/manifest/application">
            <application android:usesCleartextTraffic="true" />
        </edit-config>
```
but I don't know why it's a conflict

**Comment by liuxiaoy** [comment]
it is known as a bug of Cordova cli.[apache/cordova-common#89](https://github.com/apache/cordova-common/pull/89)

**Comment by cleverappdesign** [comment]
Still having this issue, anyone found a solution yet?

**Comment by cleverappdesign** [comment]
> Yes I still had the problem today. It works after I add this permission
> 
> ```
>  <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
> ```

Where did you add the permission?

**Comment by cleverappdesign** [comment]
### The gallery works after I added the following permission in the config.xml:
```
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```

### However, the camera not working:

`Attempt to invoke virtual method 'android.content.res.XmlResourceParser android.content.pm.PackageItemInfo.loadXmlMetaData(android.content.pm.PackageManager, java.lang.String)' on a null object reference`

At first, the initial try on opening the camera asked the user for permission, then it crashes the app. When you re-enter the app and tried to open the camera again, it throws the above error, which is same as @GuiF007 's issue.

### Adding the following code suggested aboveto config.xml did not fix my issue:
```
<config-file mode="merge" parent="application" target="AndroidManifest.xml">
    <provider android:authorities="${applicationId}.provider" android:exported="false" android:grantUriPermissions="true" android:name="org.apache.cordova.camera.FileProvider">
        <meta-data android:name="android.support.FILE_PROVIDER_PATHS" android:resource="@xml/camera_provider_paths" />
    </provider>
</config-file>
```

### Here's my ionic info
```
Ionic:
   ionic (Ionic CLI)  : 4.10.2
   Ionic Framework    : ionic-angular 3.9.3
   @ionic/app-scripts : 3.2.2

Cordova:
   cordova (Cordova CLI) : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms     : android 8.1.0, ios 5.0.1
```

Has anyone found a fix for this yet? Thanks for your help.

**Comment by thmclellan** [comment]
@cleverappdesign To get the camera working without permission issues on Android, one approach that can work is to handle permissions through a separate plugin like [cordova-plugin-android-permissions](cordova-plugin-android-permissions). 

There's an Ionic native wrapper at https://ionicframework.com/docs/native/android-permissions with an example of how to checkPermission and requestPermission. 

One hassle with troubleshooting Cordova is that the above config.xml entries describe what should get added to your AndroidManifest.xml. It may be worth checking your /platforms/android/app/src/main/AndroidManifest.xml file to see if it's getting updated with each cordova build. Hope that helps

**Comment by cleverappdesign** [comment]
> @cleverappdesign To get the camera working without permission issues on Android, one approach that can work is to handle permissions through a separate plugin like [cordova-plugin-android-permissions](cordova-plugin-android-permissions).
> 
> There's an Ionic native wrapper at https://ionicframework.com/docs/native/android-permissions with an example of how to checkPermission and requestPermission.
> 
> One hassle with troubleshooting Cordova is that the above config.xml entries describe what should get added to your AndroidManifest.xml. It may be worth checking your /platforms/android/app/src/main/AndroidManifest.xml file to see if it's getting updated with each cordova build. Hope that helps

Thank for your reply, let me try again and get back to you!

**Comment by cleverappdesign** [comment]
> /platforms/android/app/src/main/AndroidManifest.xml

Thanks for your help, after added the following permission plug-in it works. https://ionicframework.com/docs/native/android-permissions 

Tested on both simulator and the real device.

**Comment by tinoteam** [comment]
I add Filepath plugin and work fine

**Comment by jcesarmobile** [comment]
Closing since it was a bug in cordova-common and was fixed and released long ago

**Comment by AbdallaHawwas** [comment]
> I add Filepath plugin and work fine

Where did you add the file path ?
